samza-core/src/main/java/org/apache/samza/operators/impl/store/TimeSeriesStoreImpl.java

```diff
@@ -0,0 +1,185 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+package org.apache.samza.operators.impl.store;
+
+import org.apache.samza.storage.kv.ClosableIterator;
+import org.apache.samza.storage.kv.Entry;
+import org.apache.samza.storage.kv.KeyValueIterator;
+import org.apache.samza.storage.kv.KeyValueStore;
+
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Provides a view on top of a {@link KeyValueStore} that allows retrieval of entries by time ranges.
+ *
+ * <p> A {@link TimeSeriesStoreImpl} can be backed by persistent stores like rocksDB, in-memory stores, change-logged
+ * stores, cached stores (or any combination of these).
+ *
+ * <p> Range iterators in the store return values in the order of their timestamp. Within the same key and timestamp,
+ * values are returned in their order of insertion.
+ *
+ * <p> This store has two modes of operation depending on how duplicates are handled:
+ * <ol>
+ *   <li>
+ *     Overwrite Mode: In this mode, the store only retains the most recent value for a given key and timestamp. ie.,Calling
+ *     {@link #put} on an existing key and timestamp will overwrite the previously stored value for that key.
+ *   </li>
+ *   <li>
+ *     Append Mode: In this mode, the store retains all previous values for a given key and timestamp. ie., Calling {@link #put}
+ *     with an existing key and timestamp will append the value to the list.
+ *   </li>
+ * </ol>
+ * <p> Implementation Notes:
+ *
+ *  Data is serialized and organized into K-V pairs as follows:
+ *  <pre>
+ *    +-----------------------+------------------+------------+------------------------+
+ *    |  serialized-key bytes |  timestamp       | seq num    |serialized-value bytes  |
+ *    |                       |                  |            |                        |
+ *    +-----------------------+------------------+------------+------------------------+
+ *    +----------------------+--------8 bytes----+----4 bytes--
+ *    +------------------STORE KEY----------------------------+-------STORE VAL--------+
+ *  </pre>
+ *  An 8 byte timestamp, and a 4 byte sequence number are appended to the provided key and this
+ *  combination is used as the key in the k-v store. The provided value is stored as is.
+ *
+ *  <p> This class is thread-safe and concurrent reads/writes are expected.
+ *
+ * @param <K> the type of key in the store
+ * @param <V> the type of value in the store
+ */
+
+public class TimeSeriesStoreImpl<K, V> implements TimeSeriesStore<K, V> {
+
+  private final KeyValueStore<TimeSeriesKey<K>, V> kvStore;
+
+  /**
+   * Since timestamps are at the granularity of milliseconds, multiple entries added in the same
+   * millisecond are distinguished by a monotonically increasing sequence number.
+   */
+  private int seqNum = 0;
```

Comment:
I think it's reasonable to use an AtomicInteger here instead of an int + lock: 
The concern is ordering of values after the seq num wraps. We get 2G seq nums before we wrap. Given that we have 8 byte timestamp + 4 byte seqNum overhead per key, this is at the very minimum (12 bytes * 2 G) = 24 GB for a store with empty keys and values. Even an (int, int) key value store would have 8 * 24 = 192 GB storage at a minimum (there's additional rocksDB ttl timestamp and write amplification overhead), and this is per TaskInstance.

Just throwing it out there since we were also considering storing a version number with the key: we can use 7 LSB from a long seq num + a 1 byte version num.

Reply:
Thank you for the suggestion.

After discussion we agreed the following:
- The wrapping is a non-issue if we use an atomic long
- We'll use a version for the MSB, and use the next 7 bytes for storing the long value in the sequence number